src/Data/Lens/Index.purs

```diff
@@ -76,6 +78,14 @@ instance indexNonEmptyArray :: Index (NEA.NonEmptyArray a) Int a where
           (pure xs)
           (coalg >>> map \x -> fromMaybe xs (NEA.updateAt n x xs))
 
+instance indexList :: Index (List a) Int a where
+  ix n =
+    wander \coalg xs ->
+      xs L.!! n #
+        maybe
+          (pure xs)
+          (coalg >>> map \x -> fromMaybe xs (L.updateAt n x xs))
```

Comment:
This will walk the list `xs` twice: once for looking up, and once when updating.
See the haskell version for one approach to doing this in a single pass.

Reply:
@LiamGoodacre you're referring [to this class](http://hackage.haskell.org/package/lens-4.17.1/docs/src/Control.Lens.At.html#Ixed) and this instance?

```hs
instance Ixed [a] where
  ix k f xs0 | k < 0     = pure xs0
             | otherwise = go xs0 k where
    go [] _ = pure []
    go (a:as) 0 = f a <&> (:as)
    go (a:as) i = (a:) <$> (go as $! i - 1)
  {-# INLINE ix #-}
```

Reply:
Should a similar performance update be made for `Array`? The Haskell implementation for a vector is:

```hs
instance Ixed (Vector.Vector a) where
  ix i f v
    | 0 <= i && i < Vector.length v = f (v Vector.! i) <&> \a -> v Vector.// [(i, a)]
    | otherwise                     = pure v
```

Reply:
@thomashoneyman To check the length of the `List` will incur a traversal of the list and wouldn't be anymore efficient.

However, I expect that Array could be optimized but that probably should be done in another PR.

Reply:
@cscalfani I meant the first instance, `Ixed [a]`, to refer to your `List` implementation here, and for the `Ixed (Vector a)` instance to refer to `Array`. 

Does the first instance (which doesn't check the length) make sense as an optimization?

I agree that any array optimization should be done separately.

Reply:
Okay, I made the changes to mirror the Haskell code.

Note that I had to add the explicit Type Signature for the `go` function because the compiler couldn't infer the correct type for `pure Nil` for some reason.

Reply:
After looking at `indexArray`, I realized that `xs A.!! n` does not traverse the list.

Here's the implementation of `!!` an alias for `index` which is a FFI to `indexImpl`:

```javascript
exports.indexImpl = function (just) {
  return function (nothing) {
    return function (xs) {
      return function (i) {
        return i < 0 || i >= xs.length ? nothing :  just(xs[i]);
      };
    };
  };
};
```

So I don't think there is a need to rewrite `indexArray`.